### ISSUE TITLE
[FEATURE] Support advanced scheduling of VMs with affinity rules(Harvester & RKE2 only)

### ISSUE THREAD

**Issue by janeczku** [harvester/harvester#2317]
**Is your feature request related to a problem? Please describe.**

Advanced scheduling is necessary so that users may:

- configure workloads for improved availability, e.g. by running instances in in multiple zone
- ensure that workloads are scheduled to specific hardware flavors

 **Describe the solution you'd like**

Users can define advanced scheduling rules for individual VMs and guest cluster nodes.

**Expected behaviour**

`vmAntiAffinity`, `vmAffinity`, `nodeAffinity` and `nodeAntiAffinity` rules can be specified when:

- Provisioning individual VMs
- Configuring machine pools for Harvester guest clusters.

**References**

- https://kubevirt.io/user-guide/operations/node_assignment/#affinity-and-anti-affinity
- https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity

**Comment by futuretea** [comment]
Currently, we supported `nodeAffinity` and `nodeSelector` for individual VMs
and supported `nodeAffinity` for guest cluster nodes

For harvester backend and node driver backend, `vmAntiAffinity`, `vmAffinity`, `nodeAffinity` and `nodeAntiAffinity` are all supported, need UI support, cc @guangbochen @DaiYuzeng

**Comment by guangbochen** [comment]
Due to timeline constraints, we have tentatively moved the vmAntiAffinity UI future to v1.1.2, thanks.

**Comment by irishgordo** [comment]
@futuretea  - Should [terraform provider, rancher2](https://github.com/rancher/terraform-provider-rancher2) be used within the testing scope / plan of this?

**Comment by futuretea** [comment]
@irishgordo In theory, yes. rancher2 provider v3.0.0 should support this feature by `vm_affinity` field (this field corresponds to vm's `spec.template.spec.affinity`) refer to https://kubevirt.io/user-guide/operations/node_assignment/#affinity-and-anti-affinity

**Comment by TachunLin** [comment]
### Test Plans 

### Basic Test Plan
* PodAffinity
* podAntiAffinity
* conflict podAffinity <-> nodeAffinity
* conflict podAntiAffinity <-> nodeAffinity
* Negative - Empty podAffinity

#### VM Scheduling on Harvester cluster nodes 
On Harvester cluster (UI)

1. Basic Test Plans
2. Advanced
    * vm affinity and anti-affinity works on give topology keys of the following combinations
       - The VM's namespace
       - Workloads in these namespace
       - Weight
    * The multiple rules works on affinity and anti-affinity 
    * The created node/pod affinity should be found in yaml
    * Modify the configuration of node/pod affinity can be found in yaml 
    * Multiple VM scheduling apply priority

#### VM Scheduling on Harvester guest cluster machine pools
RKE2 guest cluster provisioned from external Rancher instance (UI)

1. Basic Test Plans
2. Advanced
    * vm affinity and anti-affinity works on give topology keys of the following combinations
       - The VM's namespace
       - Workloads in these namespace
       - Weight
    * The multiple rules works on affinity and anti-affinity 
    * The created node/pod affinity should be found in yaml
    * Modify the configuration of node/pod affinity can be found in yaml 
    * Multiple VM scheduling apply priority

#### Rancher2 provider
Terraform rancher 2.0 provisioned RKE2 guest cluster
1. Basic Test Plans

**Comment by TachunLin** [comment]
@DaiYuzeng , @futuretea 
After discussion, we plan to discuss the test plan and validate feature after rc1 to ensure the upstream pr merged. 
Move the issue to Review stage.

**Comment by futuretea** [comment]
# Test Plan Reference

## Env  Setup
1. Prepare a two nodes Harvester cluster. e.g. `node1` and `node2` (master version)
1. Add label `topology.kubernetes.io/zone=zone1` to `node1` and label `topology.kubernetes.io/zone=zone2` to `node2`
1. Create VM `test`, specify node affinity: `topology.kubernetes.io/zone in [zone1]`
![image](https://github.com/harvester/harvester/assets/15064560/b7b69784-3fb3-4bf6-827a-32fd521876e9)
1. The VM `test` is successfully scheduled and is scheduled to the node of `zone1`
1. Import the Harvester cluster to Rancher `v2.7-head`

## Cases
**Case 1**
1. create a vm with `podAffinity`
```json
{
  "podAffinity": {
    "requiredDuringSchedulingIgnoredDuringExecution": [
      {
        "labelSelector": {
          "matchExpressions": [
            {
              "key": "harvesterhci.io/vmName",
              "operator": "In",
              "values": [
                "test"
              ]
            }
          ]
        },
        "topologyKey": "topology.kubernetes.io/zone"
      }
    ]
  }
}
```
2. The  VM is successfully scheduled to the same zone as VM `test`

**Case 2**
1. create a vm with `podAntiAffinity`
```json
{
  "podAntiAffinity": {
    "requiredDuringSchedulingIgnoredDuringExecution": [
      {
        "labelSelector": {
          "matchExpressions": [
            {
              "key": "harvesterhci.io/vmName",
              "operator": "In",
              "values": [
                "test"
              ]
            }
          ]
        },
        "topologyKey": "topology.kubernetes.io/zone"
      }
    ]
  }
}
```
2. The VM  is successfully scheduled and will `not` be scheduled to the same zone as VM `test`

**Case 3**
1. create a vm with conflict `podAffinity` and `nodeAffinity`
```json
{
  "podAffinity": {
    "requiredDuringSchedulingIgnoredDuringExecution": [
      {
        "labelSelector": {
          "matchExpressions": [
            {
              "key": "harvesterhci.io/vmName",
              "operator": "In",
              "values": [
                "test"
              ]
            }
          ]
        },
        "topologyKey": "topology.kubernetes.io/zone"
      }
    ]
  },
  "nodeAffinity": {
    "requiredDuringSchedulingIgnoredDuringExecution": {
      "nodeSelectorTerms": [
        {
          "matchExpressions": [
            {
              "key": "topology.kubernetes.io/zone",
              "operator": "In",
              "values": [
                "zone2"
              ]
            }
          ]
        }
      ]
    }
  }
}
```
2. The VM scheduling failed

**Case 4**
1. create a vm with conflict `podAntiAffinity` and `nodeAffinity`
```json
{
  "podAntiAffinity": {
    "requiredDuringSchedulingIgnoredDuringExecution": [
      {
        "labelSelector": {
          "matchExpressions": [
            {
              "key": "harvesterhci.io/vmName",
              "operator": "In",
              "values": [
                "test"
              ]
            }
          ]
        },
        "topologyKey": "topology.kubernetes.io/zone"
      }
    ]
  },
  "nodeAffinity": {
    "requiredDuringSchedulingIgnoredDuringExecution": {
      "nodeSelectorTerms": [
        {
          "matchExpressions": [
            {
              "key": "topology.kubernetes.io/zone",
              "operator": "In",
              "values": [
                "zone1"
              ]
            }
          ]
        }
      ]
    }
  }
}
```
2. The VM scheduling failed

**Case 5**
1.  create a vm with a no rule `pod affinity`
```json
{
  "podAffinity": {
    "requiredDuringSchedulingIgnoredDuringExecution": [
      {
        "topologyKey": "topology.kubernetes.io/zone"
      }
    ]
  }
}
```
2. The VM scheduling failed

## Terraform Template

```hcl
terraform {
  required_providers {
    rancher2 = {
      source  = "rancher/rancher2"
      version = "3.0.1"
    }
  }
}


provider "rancher2" {
  api_url    = "<>"
  access_key = "<>"
  secret_key = "<>"
  insecure   = true
}


data "rancher2_cluster_v2" "foo-harvester" {
  name = "<>"
}

# Create a new Cloud Credential for an imported Harvester cluster
resource "rancher2_cloud_credential" "foo-harvester" {
  name = "foo-harvester"
  harvester_credential_config {
    cluster_id         = data.rancher2_cluster_v2.foo-harvester.cluster_v1_id
    cluster_type       = "imported"
    kubeconfig_content = data.rancher2_cluster_v2.foo-harvester.kube_config
  }
}

data "template_file" "affinity" {
  template = <<EOF
  {
    "podAffinity": {
      "requiredDuringSchedulingIgnoredDuringExecution": [
        {
          "labelSelector": {
            "matchExpressions": [
              {
                "key": "harvesterhci.io/vmName",
                "operator": "In",
                "values": [
                  "test"
                ]
              }
            ]
          },
          "topologyKey": "topology.kubernetes.io/zone"
        }
      ]
    }
  }
  EOF
}

data "template_file" "userdata" {
  template = <<EOF
#cloud-config
package_update: true
packages:
  - qemu-guest-agent
  - iptables
runcmd:
  - - systemctl
    - enable
    - '--now'
    - qemu-guest-agent.service
EOF
}

# Create a new rancher2 machine config v2 using harvester node_driver
resource "rancher2_machine_config_v2" "foo-harvester-v2" {
  generate_name = "foo-harvester-v2"
  harvester_config {
    vm_namespace = "default"
    cpu_count    = "2"
    memory_size  = "4"
    vm_affinity  = base64encode(data.template_file.affinity.rendered)
    #vm_affinity  = data.template_file.affinity.rendered
    disk_info    = <<EOF
    {
        "disks": [{
            "imageName": "<>",
            "size": 40,
            "bootOrder": 1
        }]
    }
    EOF
    network_info = <<EOF
    {
        "interfaces": [{
            "networkName": "<>"
        }]
    }
    EOF
    ssh_user     = "ubuntu"
    user_data    = base64encode(data.template_file.userdata.rendered)
  }
}

resource "rancher2_cluster_v2" "foo-harvester-v2" {
  name               = "foo-harvester-v2"
  kubernetes_version = "v1.26.5+rke2r1"
  rke_config {
    machine_pools {
      name                         = "pool1"
      cloud_credential_secret_name = rancher2_cloud_credential.foo-harvester.id
      control_plane_role           = true
      etcd_role                    = true
      worker_role                  = true
      quantity                     = 1
      machine_config {
        kind = rancher2_machine_config_v2.foo-harvester-v2.kind
        name = rancher2_machine_config_v2.foo-harvester-v2.name
      }
    }
    machine_selector_config {
      config = {
        cloud-provider-name = ""
      }
    }
    machine_global_config = <<EOF
cni: "calico"
disable-kube-proxy: false
etcd-expose-metrics: false
EOF
    upgrade_strategy {
      control_plane_concurrency = "10%"
      worker_concurrency        = "10%"
    }
    etcd {
      snapshot_schedule_cron = "0 */5 * * *"
      snapshot_retention     = 5
    }
    chart_values = ""
  }
}

```

**Comment by TachunLin** [comment]
### Test Plan matrix

|                                      | podAffinity | podAntiAffinity | conflict  podAffinity <-> nodeAffinity | conflict  podAntiAffinity <-> nodeAffinity | Negative - Empty podAffinity |
|--------------------------------------|:-----------:|:---------------:|:--------------------------------------:|:------------------------------------------:|----------------------------|
| Multi-cluster RKE2 (Ready)           |    PASS         |    PASS      |              PASS                          |                  PASS                     |  PASS                           |
| Terraform rancher 2.0 - RKE2 (Ready) |    PASS     |     PASS        |         PASS                           |             PASS                |    PASS                                                        |
| Single-cluster  Harvester (Ready)            |   PASS      |  PASS           |         PASS                           |   PASS                        |   PASS                         |
| Multi-cluster RKE1                   |             |                 |                                        |                                            |                            |
| Terraform rancher 2.0 - RKE1         |             |                 |                                        |                                            |                            |

### Expected Result
1. PodAffinity: 
   - The VM is successfully scheduled to the same zone as VM test
  ![image](https://github.com/harvester/harvester/assets/29251855/69bcb16a-2166-4300-a14a-3ae45ddcef07)

   
2. podAntiAffinity: 
   - The VM is successfully scheduled and will not be scheduled to the same zone as VM test
    ![image](https://github.com/harvester/harvester/assets/29251855/79de8a17-bf86-4d0a-93b2-f5f5fbb746ba)

   
3. conflict  podAffinity <-> nodeAffinity
   - The VM scheduling failed with message `3 Preemption is not helpful for scheduling.`
    ![image](https://github.com/harvester/harvester/assets/29251855/3a523e18-a294-42e2-9fd0-ef1b6d40d21f)

   
4. conflict  podAntiAffinity <-> nodeAffinity
   - The VM scheduling failed with message `3 Preemption is not helpful for scheduling.`
    ![image](https://github.com/harvester/harvester/assets/29251855/c84bc442-5a9d-4e78-958e-9b3d940d3ae3)

   
5. Negative Empty podAffinity
   - The VM scheduling failed
    ![image](https://github.com/harvester/harvester/assets/29251855/8f6e6406-a39b-46cb-b7f8-84177627c094)

#### Additional  terraform-provider-rancher2 test plan (Check abnormal)

1. If users didn't edify anything, clusters should not be recreated. 
    - no edit
    - add then remove new attribute
    - remove then add new attribute
6. terraform-provider-rancher2 can pass the json format to guest cluster without pending in the loading  
  ```
  #vm_affinity  = data.template_file.affinity.rendered
  ```
### Test Information
* Test Environment: 3 nodes Harvester bare machines
* Harvester version: latest v1.2-head , v1.2.0-rc2
* Rancher version: v2.7-head
* terraform-provider-rancher2: v3.0.1

**Comment by TachunLin** [comment]
Complete the current available test plan path of:

1. Multi-cluster RKE2 integration for UI 
2. Multi-cluster RKE2 integration for Terraform rancher 2.0

Move to implement and continue to validate when other test path PR have been merged. 

1. Single cluster on Harvester
2. Multi-cluster RKE1 integration for UI
3. Multi-cluster RKE1 integration for Terraform rancher 2.0

**Comment by TachunLin** [comment]
Tested on the following test plan path: 
1. Multi-cluster vm provisioning on RKE2 instance from Rancher UI
2. Multi-cluster vm provisioning on RKE2 instance from terraform-rancher2.0
3. Single-cluster vm provisioning on Harvester

Move to ready for testing and plan to continue after the RKE1 content is ready.

**Comment by DaiYuzeng** [comment]
Hi! @TachunLin, RKE1 related changes won't be merged at this moment. I filed an issue to track https://github.com/harvester/harvester/issues/4121. Can we close this issue if there is no question and track RKE1 changes at the new issue?

**Comment by TachunLin** [comment]
Thanks @DaiYuzeng for the comment, we would track the support status of RKE1 podaffinity feature in issue #4121. 
The `Multi-cluster RKE1` and `Terraform rancher 2.0 - RKE1` test scope would not be implement in current issue.

**Comment by TachunLin** [comment]
Verified **failed** on `v1.2.0-rc2`. Suggest moving to implement for further check. 

While validating the advanced test plan, we found an potential issue. 
When we edit a running VM already have podaffinity rule set, change the `Operator` from `in list` to `not in list`. 
Save then restart the VM, we found the operator did not change accordingly on UI nor on the yaml content. 

https://github.com/harvester/harvester/assets/29251855/4d7320c1-fa97-4d5d-af43-c4cc8ad1e363

![image](https://github.com/harvester/harvester/assets/29251855/56fa52b8-4d01-4e6a-bdef-b02eb81d2c9c)

**Comment by DaiYuzeng** [comment]
Also occurred in Rancher Pod configurations. I am going to work on it. Thanks for your validation @TachunLin !

**Comment by TachunLin** [comment]
- [x] Verified the `in list` and `not in list` operator for vm affinity work as expected 
  ![image](https://github.com/harvester/harvester/assets/29251855/b303386e-9b95-4219-88a0-4a35815bea89)


  ```
  podAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              - labelSelector:
                  matchExpressions:
                    - key: harvesterhci.io/vmName
                      operator: NotIn
                      values:
                        - vm1
                topologyKey: topology.kubernetes.io/zone
  ```

- [x] Verified the `is set` and `not in set` operator for vm `affinity` and `anti-affinity` work as expected 

- [x] Verified for node affinity the `Required` rule have higher priority to be applied than `Preferred` 

- [x] Verified we **can't** have two `Required` affinity rule set on the same VM 
- [x] Verified we **can** have two `Required` affinity rule set on the same VM, it will apply in sequence

**Comment by TachunLin** [comment]
Verified **failed** on `v1.2-f4968e49-head` (6/30) since found a potential issue 
Move the issue to implement 

### Result
When we have a running vm contain two vm affinity set on it. 
Then we edit the `Weight` field with **backspace** key, the weight field will be deleted and no longer display again. 


https://github.com/harvester/harvester/assets/29251855/6af6d956-616f-402b-ad9c-8a2dd4f25e44

* And we can save this setting and found the weight have no longer exists
![image](https://github.com/harvester/harvester/assets/29251855/117d9fde-14c6-438f-aeb8-c286c5bdc904)

Suggest we can keep the weight value in empty while clicking the backspace key to the end. 
To provide a consistent user experience on the vm affinity page.
 
### Test Information
* Test Environment: 3 nodes harvester on provo bare machine
* Harvester version: `v1.2-f4968e49-head` (6/30)

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.2-bd26e791-head` (07/06) with Rancher v2.7-head (3319bfb). Close this issue.

### Result
* We can use backspace key to edit the `weight` field in vm affinity page for both `single` and `multiple` cluster scenarios
![image](https://github.com/harvester/harvester/assets/29251855/54091a21-378b-4299-bb6c-a88768485ba8)

* Also tested the `priority`, `operator` and `weight` works on **multiple** cluster environment just as the single cluster test scenarios in:  
https://github.com/harvester/harvester/issues/2317#issuecomment-1617801441

### Test Information
* Test Environment: 3 nodes harvester clusters on bare machine
* Harvester version: `v1.2-bd26e791-head` (07/06)
* Rancher: v2.7-head (3319bfb)

### Verify Steps
https://github.com/harvester/harvester/issues/2317#issuecomment-1549722809
https://github.com/harvester/harvester/issues/2317#issuecomment-1580465914
https://github.com/harvester/harvester/issues/2317#issuecomment-1589442468
